cmd/app.go

```diff
@@ -238,30 +260,93 @@ func appLs(ctx *cli.Context) error {
 	}
 
 	collection, err := c.ProjectClient.App.List(defaultListOpts(ctx))
+	if err != nil {
+		return err
+	}
+
 	writer := NewTableWriter([][]string{
 		{"ID", "ID"},
 		{"NAME", "App.Name"},
 		{"STATE", "App.State"},
+		{"CATALOG", "Catalog"},
+		{"TEMPLATE", "Template"},
 		{"VERSION", "Version"},
 	}, ctx)
 
 	defer writer.Close()
 
 	for _, item := range collection.Data {
-		parsedExternal, err := parseExternalID(item.ExternalID)
+		appData := &AppData{
+			ID:  item.ID,
+			App: item,
+		}
+
+		appExternalID := item.ExternalID
+		var appTemplateFiles map[string]string
+		if appExternalID == "" {
+			// add namespace prefix to AppRevisionID to create a Rancher API style ID
+			appRevisionID := strings.Replace(item.ID, item.Name, item.AppRevisionID, -1)
+
+			appRevision, err := c.ProjectClient.AppRevision.ByID(appRevisionID)
+			if err != nil {
+				return err
+			}
+
+			appTemplateFiles = appRevision.Status.Files
+		}
+
+		appData.Catalog, appData.Template, appData.Version, err = parseTemplateInfo(appExternalID, appTemplateFiles)
 		if err != nil {
 			return err
 		}
-		writer.Write(&AppData{
-			ID:      item.ID,
-			App:     item,
-			Version: parsedExternal["version"],
-		})
+
+		writer.Write(appData)
 	}
 	return writer.Err()
 
 }
 
+func parseTemplateInfo(appExternalID string, appTemplateFiles map[string]string) (templateCatalog, templateName, templateVersion string, err error) {
```

Comment:
Have this function return a map like `parseExternalID` and an error. Naked returns should be for short functions that are easy to follow.

Reply:
Refactored